File: packages/api-contract/src/runs.ts
import { z } from "zod";
import { channelId, run, runId, runStep } from "@perch/core";

export const getRunInput = z.object({ runId });
export const getRunOutput = z.object({
  run,
  steps: z.array(runStep),
});

/** Backs ChatScreen's live "Run #… running" card (see apps/desktop/src/App.tsx's
 * `activeRunsByChannel`) — without this, a run already in flight when the app is opened or
 * reconnects has nothing to show until its next SSE event. `channelId` is required, not optional:
 * this is a per-channel live-status query, not a general run listing (see runs.ts's own comment
 * for why it isn't indexed/paginated). */
export const listActiveRunsInput = z.object({ channelId });
export const listActiveRunsOutput = z.array(run);

/** Lets a person give up on a run that's stuck `"running"`/`"waiting_approval"` — see
 * services/api/src/routers/runs.ts's `POST /{runId}/cancel` for what this actually does (marks the
 * DB row failed; does NOT reach into AWS to stop the underlying durable execution — see that
 * route's comment for why). */
export const cancelRunInput = z.object({ runId });
export const cancelRunOutput = run;

File: apps/desktop/src-tauri/src/api/runs.rs
use reqwest::Method;
use tauri::AppHandle;

use super::client::call;
use super::types::*;

#[tauri::command]
pub async fn runs_get(app: AppHandle, run_id: String) -> Result<GetRunsRunIdResponse, String> {
    call(&app, Method::GET, &format!("/runs/{run_id}"), None::<&()>).await
}

/// The channel's runs still `"running"`/`"waiting_approval"` — backs ChatScreen's live run
/// indicator on load/reconnect (see services/api/src/routers/runs.ts's `GET /` for why this
/// exists alongside the SSE-driven `run.updated` path).
#[tauri::command]
pub async fn runs_list_active(app: AppHandle, channel_id: String) -> Result<Vec<GetRunsResponseItem>, String> {
    call(&app, Method::GET, &format!("/runs?channelId={channel_id}"), None::<&()>).await
}

/// Gives up on a stuck run — see services/api/src/routers/runs.ts's `POST /{runId}/cancel` for
/// exactly what this does and does not do (marks it failed; doesn't reach into AWS).
#[tauri::command]
pub async fn runs_cancel(app: AppHandle, run_id: String) -> Result<PostRunsRunIdCancelResponse, String> {
    call(&app, Method::POST, &format!("/runs/{run_id}/cancel"), None::<&()>).await
}
